SFLASH support and offline updater

tdz opened the pull request.

tdz commented:
FYI: For testing, I copied the blinky example to /update/mcuimg.bin. After flashing and running the updater application, the device executes blinky.

tdz commented:
@fabricedesre  Just friendly ping for a review of this PR.

tdz commented:
Updated the PR
 * Added Image::update() for selecting the image's file name automatically
 * Removed 'failsafe' parameter from Image::create(): it's not supported by the SDK

fabricedesre commented on src/io.rs:
```diff
@@ -18,10 +18,30 @@ macro_rules! try_fs {
+
+    fn read_to_string(&mut self, str: &mut String) -> Result<usize, SimpleLinkError> {
+        let mut buf: Vec<u8> = Vec::new();
+        let len = try!(self.read_to_end(&mut buf));
```
nit (here and other places): I think you can use let len = self.read_to_end(&mut buf)?;

tdz replied:
Wouldn't the compiler complain about an unused Result<>?

tdz replied:
Tried and the compiler stops with an error.

fabricedesre replied:
Which error? using the `?` operator is very similar to the `try!` macro.

tdz replied:
I see. I thought the question mark was part of the sentence. :)

fabricedesre commented on src/io.rs:
```diff
@@ -138,34 +162,156 @@ impl Read for File {
+
+//
+// Image files
+//
```
Can we move all this code in it's own file?

tdz replied:
Image belongs into the io namespace. Can namespaces be split up into multiple files (e.g., io_traits.rs, io_file.rs, io_image.rs)?

tdz replied:
The interface for testing and committing images is a feature of the boot loader and rather unrelated to reading and writing image files. Therefore I moved these methods from io::Image to cc3200::Update. That should reflect the board-wide nature of these operations.

Being about file access, I still think that Image belongs into io. Internally in the SDK it's mostly the same code as for regular files with some extras for opening. If you really want to see Image being gone, I guess we could merge Image into File without a loss of functionality.

fabricedesre replied:
No we can't split namespaces afaik. But I'm fine with having the higher level functions in cc3200::Update

fabricedesre commented on src/io.rs:
```diff
@@ -138,34 +162,156 @@ impl Read for File {
+
+pub struct Image {
+    offset: usize,
+    file_handle: i32,
```
for a followup: we should consider not using i32 for the type here.

tdz replied:
Why? This is the type that the underlying C implementation uses.

fabricedesre replied:
Sure, but just declaring a type alias like:
```rust
type FileHandle = i32;
```
will hide the implementation details. No big deal anyway.

fabricedesre commented on src/io.rs:
```diff
@@ -138,34 +162,156 @@ impl Read for File {
+    /// Opens the next image for updating; the filename is selected automatically
+    pub fn update(max_len: usize) -> Result<Image, SimpleLinkError> {
+        let next_image_name : &'static str = "/sys/mcuimgA.bin"; // selects next image
+        Image::create(next_image_name, max_len)
```
No need to declare a static str here: Image::create("/sys/mcuImgA.bin", max_len)

tdz replied:
It's for that comment next to the constant.

tdz replied:
I put the comment in front of the create() statement.

fabricedesre commented on examples/update.rs:
```diff
@@ -0,0 +1,147 @@
+
+    // BUG? tzimmermann: I got a FILE_NOT_ERROR if I reverse
+    // the next two lines.
+
```
is that happening when doing File::get_info(filename) ?

tdz replied:
Not sure, but I think so.

tdz replied:
This has been fixed in the latest PR.

tdz commented:
I pushed a new patch set into the PR that updates as explained above.

tdz commented:
Updated the PR

  * replaced try! by ?

tdz commented:
Updated the PR:

  - fixed string conversion rust<>C, which makes file-name handling reliable
  - removed Image data structure; now implemented as File
  - added examples simple-ota, full-ota, and full-ota-mcuimga that demonstrate OTA mechanisms

tdz commented:
@dhylands Dave, can you continue with the review of this patch set? I'd like to get it merged ASAP.

fabricedesre commented on examples/full-ota.rs:
```diff
@@ -0,0 +1,371 @@
+            ignore!(File::remove(UPDATE_IMG));
+            return Err(err);
+        },
+    };
```
that could be:
```rust
if let Err(err) = get_update(...) {
    ignore!(File::remove(...));
    return Err(err);
}

(same change can be done in other examples)

fabricedesre commented on examples/simple-ota.rs:
```diff
@@ -0,0 +1,367 @@
+            Ok(_) => { },
+            Err(_) => { },
+        }
+    })
```
maybe the match can be simplified to:
```rust
match $e {
    _ => { },
}
```

tdz commented:
Updated PR with the suggested changes.
